node/pkg/cni/token_watch.go

```diff
@@ -1,56 +1,245 @@
+	var nextExpiration time.Time
+	rand := rand.New(rand.NewSource(time.Now().UnixNano()))
 	for {
-		// To prevent tight looping, add a sleep here.
-		time.Sleep(1 * time.Second)
+		tu, err := t.UpdateToken()
+		if err != nil {
+			logrus.WithError(err).Error("Failed to update CNI token, retrying...")
+			// Reset nextExpiration to retry directly
+			nextExpiration = time.Time{}
+		} else {
+			nextExpiration = tu.ExpirationTime
+			t.tokenChan <- tu
+		}
+		now := time.Now()
+		var sleepTime time.Duration
+		// Do some basic rate limiting to prevent flooding the kube apiserver with requests
+		if nextExpiration.Before(now.Add(t.minTokenRetryDuration * t.defaultRefreshFraction)) {
+			sleepTime = t.minTokenRetryDuration
+		} else {
+			sleepTime = nextExpiration.Sub(now) / t.defaultRefreshFraction
+		}
+		jitter := rand.Float32() * float32(sleepTime)
+		sleepTime += time.Duration(jitter)
+		if logrus.IsLevelEnabled(logrus.DebugLevel) {
+			logrus.Debugf("Going to sleep for %s", sleepTime.String())
+		}
+		time.Sleep(sleepTime)
+	}
+}
+
+func (t *TokenRefresher) tokenRequestSupported(clientset *kubernetes.Clientset) bool {
+	t.tokenOnce.Do(func() {
+		resources, err := clientset.Discovery().ServerResourcesForGroupVersion("v1")
+		if err != nil {
+			return
+		}
+		for _, resource := range resources.APIResources {
+			if resource.Name == "serviceaccounts/token" {
+				t.tokenSupported = true
+				return
+			}
+		}
+	})
+	return t.tokenSupported
+}
+
+func tokenUpdateError(err error) (TokenUpdate, error) {
+	return TokenUpdate{}, err
+}
+
+func tokenUpdateFromFile() (TokenUpdate, error) {
+	tokenBytes, err := ioutil.ReadFile(tokenFile)
+	if err != nil {
+		logrus.WithError(err).Error("Failed to read service account token file")
+		return tokenUpdateError(err)
+	}
+	token := string(tokenBytes)
+	tokenSegments := strings.Split(token, ".")
+	if len(tokenSegments) != 3 {
+		err := fmt.Errorf("invalid token segment size: %d", len(tokenSegments))
+		logrus.WithError(err).Error("Failed parsing service account token")
+		return tokenUpdateError(err)
+	}
+	unparsedClaims := tokenSegments[1]
+	// Padding may be missing, hence check and append it
+	if l := len(unparsedClaims) % 4; l > 0 {
+		unparsedClaims += strings.Repeat("=", 4-l)
+	}
+	decodedClaims, err := base64.URLEncoding.DecodeString(unparsedClaims)
+	if err != nil {
+		logrus.WithError(err).Error("Failed to decode service account token claims")
+		return tokenUpdateError(err)
+	}
+	var claimMap map[string]interface{}
+	err = json.Unmarshal(decodedClaims, &claimMap)
+	if err != nil {
+		logrus.WithError(err).Error("Failed to unmarshal service account token claims")
+		return tokenUpdateError(err)
+	}
+	return TokenUpdate{
+		Token:          token,
+		ExpirationTime: time.Unix(int64(claimMap["exp"].(float64)), 0),
+	}, nil
+}
 
+func Run() {
+	// Log to stdout.  this prevents our logs from being interpreted as errors by, for example,
+	// fluentd's default configuration.
+	logrus.SetOutput(os.Stdout)
```

Comment:
Let's actually delete this line - since we call Run() from the test code, this will mess up our test logging and emit logs even for successful tests.

I think this is already redundant, because we set the log output properly in `main.go` already.  https://github.com/projectcalico/calico/blob/9687764b6e9a8ebc0592e3c811b4d8a100fd7b6a/node/cmd/calico-node/main.go#L87-L89

Reply:
Done.